ClangSharpPInvokeGenerator/CursorWriter.cs

```diff
@@ -41,1727 +34,1262 @@ public void Dispose()
 
-                        case "int16_t":
-                        {
-                            return "short";
-                        }
+                            if (_config.GenerateUnsafeCode)
+                            {
+                                name = GetTypeName(decl, pointeeType);
+                                name += '*';
+                            }
+                            else
+                            {
+                                _outputBuilder.AddUsing("System");
+                            }
 
-                        case "int32_t":
-                        {
-                            return "int";
+                            return name;
                         }
 
-                        case "int64_t":
+                        case CXCursorKind.CXCursor_ParmDecl:
                         {
-                            return "long";
-                        }
+                            var name = GetTypeName(decl, pointeeType);
 
-                        case "intptr_t":
-                        {
-                            _outputBuilder.AddUsing("System");
-                            return "IntPtr";
+                            if (_config.GenerateUnsafeCode)
+                            {
+                                name += '*';
+                            }
+                            return name;
                         }
 
-                        case "size_t":
-                        case "SIZE_T":
+                        case CXCursorKind.CXCursor_TypedefDecl:
                         {
-                            _outputBuilder.AddUsing("System");
-                            return "IntPtr";
-                        }
+                            var typedefDecl = (TypedefDecl)decl;
+                            var underlyingType = typedefDecl.UnderlyingType;
 
-                        case "time_t":
-                        {
-                            return "long";
-                        }
+                            if ((underlyingType.Kind == CXTypeKind.CXType_Pointer) && (underlyingType.PointeeType.Kind == CXTypeKind.CXType_FunctionProto))
+                            {
+                                goto case CXCursorKind.CXCursor_FunctionDecl;
+                            }
 
-                        case "uint8_t":
-                        {
-                            return "byte";
+                            var name = GetCursorName(pointeeType.DeclarationCursor);
+
+                            if (_config.GenerateUnsafeCode)
+                            {
+                                name += '*';
+                            }
+                            return name;
                         }
 
-                        case "uint16_t":
+                        default:
                         {
-                            return "ushort";
+                            Unhandled(decl, pointeeType);
+       
```

Comment:
This handles nested declarations, if they exist and places them just after the list of fields.